portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -32,44 +39,84 @@
 public class DataServlet extends HttpServlet {
 
   private HashMap<String, HashMap<String,Object>> information;  
-  private List<String> messages;
+
+  private String regex = "^[a-zA-Z ]+$";
+  private Pattern pattern =  Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
+  private Matcher matcher = pattern.matcher(""); 
 
   @Override
   public void init(){
-      messages = new ArrayList<>();
       information = new HashMap<String, HashMap<String,Object>>();
+      information.put("messageInfo", new HashMap<String, Object>());
+      information.put("nameInfo", new HashMap<String, Object>());
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     
+        // Query comments from DataStore as entities
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity: results.asIterable()) {
+        long id = entity.getKey().getId();
+        String name = (String) entity.getProperty("name");
+        String message = (String) entity.getProperty("message");
+        long timestamp = (long) entity.getProperty("timestamp");
+
+        Comment comment = new Comment(id, name, message, timestamp);
+        comments.add(comment);
+    }
+
+    information.get("messageInfo").put("history", comments);
+
     String json = convertToJsonUsingGson(information);
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException{
-      
+
       information = checkValidity(request);
 
+          // Send invalid input to client for correction
       if (((Boolean) information.get("nameInfo").get("error") == true) || 
           ((Boolean) information.get("messageInfo").get("error") == true)){
-          information.get("messageInfo").put("history", messages);
+
           response.sendRedirect("/index.html#comment-page");
       } else {
+          // Store valid input in datastore
           String message = (String) information.get("messageInfo").get("message");
-          messages.add(message);
-          information.get("messageInfo").put("history", messages);
+          String name = (String) information.get("nameInfo").get("name");
+
+          createEntity("Comment", name, message);
+
           response.sendRedirect("/index.html#comment-page");
```

Comment:
seems like both the if and else redirect to the same place. If so, maybe move it after the if/else block